src/Workspaces/Remote/Core/ProjectSystem/IWorkspaceProject.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using StreamJsonRpc;
+
+namespace Microsoft.CodeAnalysis.Remote.ProjectSystem
+{
+    [RpcMarshalable]
+    public interface IWorkspaceProject : IAsyncDisposable
```

Comment:
How do we feel about the chattiness of this API, now it's moving to RPC?

My napkin math suggests we're introducing 5-6 IPC calls per target per project on solution load.

Reply:
something to pay attention to.  especially if roslyn itself is like 600 projects.

Reply:
My assumption was this wouldn't be the biggest deal:

1. Latency for an RPC should be fairly cheap (this isn't doing RPC across a network.)
2. For a single project, yes, there's a bit of back and forth.
3. For larger solutions, there will be a number of projects loading all in parallel, so I'd expect _individual_ request latency won't really matter -- the pipe will be full, or if it's _not_ it's because one request is actually being processed under one of the larger workspace locks that will have to exist.

We could reduce the amount of RPC but if that's still sending all the same data over the wire if we were to make a single call with all the information at once.

There's one potential way to reduce the amount of info going across the wire which would be to only process the command line string and nothing further. Right now even in the existing in-proc system there is some redundancy: we send over the command line string so we can have compiler switches, but that _also_ includes the list of source files and references as full paths. But we let the project system directly add source files/references during earlier stages of the load when it can potentially use approximate answers from the evaluation pass before we have full answers. There's a number of ways we can adjust this, depends on various goals we want to aim for.

Reply:
Seems like it'd be better to have a single RPC for each project interaction -- i.e. group all info that project system has at certain point in time. One call can be passing command line and another a set of all items (metadata, source, additional, etc. together) that are available.

Reply:
When I brought this up with @jacdavis his thought was that this level of chatiness was unlikely to be a problem in practice, and even if it is we have ways of dealing with it.

My preference here is to get something in and working end-to-end and then iterate on the design as needed. On the project system side we can also look into adding telemetry to measure how long it is actually taking to push the information through to Roslyn, rather than speculating.